tls/s2n_signature_algorithms.c

```diff
@@ -107,12 +107,28 @@ static int s2n_tls13_default_sig_scheme(struct s2n_connection *conn, struct s2n_
     GUARD(s2n_connection_get_signature_preferences(conn, &signature_preferences));
     notnull_check(signature_preferences);
 
-    struct s2n_cipher_suite *cipher_suite = conn->secure.cipher_suite;
-    notnull_check(cipher_suite);
+    if (conn->mode == S2N_SERVER) {
+        struct s2n_cipher_suite *cipher_suite = conn->secure.cipher_suite;
+        notnull_check(cipher_suite);
+    }
 
     for (size_t i = 0; i < signature_preferences->count; i++) {
         const struct s2n_signature_scheme *candidate = signature_preferences->signature_schemes[i];
 
+        /* From RFC https://tools.ietf.org/html/rfc8446#section-4.4.3:
+         * For TLS1.3 RSA signatures MUST use an RSASSA-PSS algorithm, regardless of whether RSASSA-PKCS1-v1_5 
+         * algorithms appear in "signature_algorithms". The SHA-1 algorithm MUST NOT be used in any signatures of
+         * CertificateVerify messages. All SHA-1 signature algorithms in this specification are defined
+         * solely for use in legacy certificates and are not valid for CertificateVerify signatures.
+         * */
+        if (candidate->hash_alg == S2N_HASH_SHA1) {
+            continue;
+        }
+
+        if (candidate->sig_alg == S2N_SIGNATURE_RSA) {
+           continue;
+        }
```

Comment:
Why aren't these checks covered by s2n_is_signature_scheme_usable?

Reply:
These checks are required only for TLS1.3 versions and `s2n_is_signature_scheme_usable` is called for <= TLS1.2 versions as well. We could move these checks within `s2n_is_signature_scheme_usable` and validate under an if-condition as an option, but imo these checks being within `s2n_tls13_default_sig_scheme` logic  flow made more sense.

Reply:
My point is that these checks aren't only needed when choosing defaults. They're needed when choosing any signature scheme. 

I think they're currently covered by the maximum_protocol_version field of signature schemes, which is checked by `s2n_is_signature_scheme_usable`.